app/controllers/application_controller.rb

```diff
@@ -5,6 +5,169 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    #added below
+    enable :sessions
+    set :session_secret, "hashtagwars"
   end
 
+  get '/' do 
+    erb :index
+  end
+
+  get '/login' do 
+    if is_logged_in?
+      redirect '/tweets'
+    else
+      erb :"users/login"
+    end
+  end
+
+  post '/login' do
+    @user = User.find_by_username(params[:username])
+    session[:id] = @user[:id]
+    if @user.password == params[:password]
+      redirect '/tweets'
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/signup' do
+    if is_logged_in?
+      redirect '/tweets'
+    else
+      erb :"users/create_user"
+    end
+  end
+
+  post '/signup' do
+    if params[:username] == "" || params[:email] == "" || params[:password] == ""
+      redirect '/signup'
+    end
+    @user = User.find_by(username: params[:username])
+    if @user == nil
+      @user = User.create(username: params[:username], email: params[:email], password: params[:password])
+      @user.save
+
+      session[:id] = @user.id
+
+      redirect '/tweets'
+    elsif @user.password == params[:password]
+      session[:id] = @user.id
+      redirect '/tweets'
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/tweets' do
```

Comment:
This controller is overloaded. Consider separating your concerns by splitting some of these actions into separate controllers (i.e. a users controller and a tweets controller).